tachyon/crypto/commitments/merkle_tree/binary_merkle_tree/binary_merkle_tree.h

```diff
@@ -0,0 +1,168 @@
+#ifndef TACHYON_CRYPTO_COMMITMENTS_MERKLE_TREE_BINARY_MERKLE_TREE_BINARY_MERKLE_TREE_H_
+#define TACHYON_CRYPTO_COMMITMENTS_MERKLE_TREE_BINARY_MERKLE_TREE_BINARY_MERKLE_TREE_H_
+
+#include <algorithm>
+#include <utility>
+#include <vector>
+
+#include "gtest/gtest_prod.h"
+
+#include "tachyon/base/bits.h"
+#include "tachyon/base/logging.h"
+#include "tachyon/base/numerics/checked_math.h"
+#include "tachyon/base/openmp_util.h"
+#include "tachyon/base/range.h"
+#include "tachyon/crypto/commitments/merkle_tree/binary_merkle_tree/binary_merkle_hasher.h"
+#include "tachyon/crypto/commitments/merkle_tree/binary_merkle_tree/binary_merkle_proof.h"
+#include "tachyon/crypto/commitments/merkle_tree/binary_merkle_tree/binary_merkle_tree_storage.h"
+#include "tachyon/crypto/commitments/vector_commitment_scheme.h"
+
+namespace tachyon::crypto {
+
+template <typename LeafTy, typename HashTy, size_t MaxSize>
+class BinaryMerkleTree
+    : public VectorCommitmentScheme<BinaryMerkleTree<LeafTy, HashTy, MaxSize>> {
+ public:
+  constexpr static size_t kDefaultLeavesSizeForParallelization = 1024;
+
+  BinaryMerkleTree() = default;
+  BinaryMerkleTree(BinaryMerkleTreeStorage<HashTy>* storage,
+                   BinaryMerkleHasher<LeafTy, HashTy>* hasher)
+      : storage_(storage), hasher_(hasher) {}
+
+  size_t leaves_size_for_parallelization() const {
+    return leaves_size_for_parallelization_;
+  }
+  void set_leaves_size_for_parallelization(
+      size_t leaves_size_for_parallelization) {
+    CHECK(base::bits::IsPowerOfTwo(leaves_size_for_parallelization));
+    leaves_size_for_parallelization_ = leaves_size_for_parallelization;
+  }
+
+  // VectorCommitmentScheme methods
+  size_t N() const { return MaxSize; }
+
+  template <typename ContainerTy>
+  [[nodiscard]] bool Commit(const ContainerTy& leaves, HashTy* out) const {
+    if (!FillLeaves(leaves)) return false;
+
+    // For instance, if |leaves_size_for_parallelization_| equals 4, the
+    // subtrees with root indices 1 and 2 will be constructed.
+    //
+    //         0
+    //    1          2
+    //  3   4     5    6
+    // 7 8 9 10 11 12 13 14
+    //
+    // Finally, the remaining tree should be constructed from leaves 1 and 2.
+    size_t leaves_size = std::size(leaves);
+    OPENMP_PARALLEL_FOR(size_t i = 0; i < leaves_size;
+                        i += leaves_size_for_parallelization_) {
+      size_t from = leaves_size - 1 + i;
+      size_t to = from + leaves_size_for_parallelization_;
+      BuildTreeFromLeaves(base::Range<size_t>(from, to));
+    }
+    if (leaves_size > leaves_size_for_parallelization_) {
+      size_t i = base::bits::Log2Floor(leaves_size) -
+                 base::bits::Log2Floor(leaves_size_for_parallelization_);
+      BuildTreeFromLeaves(
+          base::Range<size_t>((1 << i) - 1, (1 << (i + 1)) - 1));
+    }
+    *out = storage_->GetHash(0);
+    return true;
+  }
+
+  [[nodiscard]] bool DoCreateOpeningProof(
```

Comment:
Is there any reason to make `DoCreateOpeningProof` and `DoVerifyOpeningProof` public? I think it would be better to make it private for consistency.